pkg/kubelet/images/image_manager.go

```diff
@@ -154,8 +157,22 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
 	if m.backOff.IsInBackOffSinceUpdate(backOffKey, m.backOff.Clock.Now()) {
 		msg := fmt.Sprintf("Back-off pulling image %q", imgRef)
 		m.logIt(objRef, v1.EventTypeNormal, events.BackOffPullImage, logPrefix, msg, klog.Info)
+
+		// Use the error from the actual pull instead of the generic
+		// ErrImagePullBackOff. This information is populated to the pods
+		// .status.containerStatuses[*].state.waiting.reason and is also used
+		// by other API consumers like the `kubectl get pod` STATUS output
+		// field.
+		prevPullErr, ok := m.prevPullErr.Load(backOffKey)
+		if ok {
+			return "", msg, prevPullErr.(error)
+		}
+
 		return "", msg, ErrImagePullBackOff
```

Comment:
So realistically we will never see `ErrImagePullBackOff` any longer?

Reply:
So this logic will never trigger: https://github.com/kubernetes/kubernetes/blob/e5ba5cd2b0d3bddd4a0393b98fde14960f31a33c/pkg/kubelet/kubelet.go#L1968-L1972

I wonder if we have any more dependencies internally or externally on this error message

Reply:
I changed the logic to work only for both known CRI errors `RegistryUnavailable` and `SignatureValidationFailed`. Generic errors will still toggle between `ImagePullBackOff` and `ErrImagePull`.

Would that work?

Reply:
I know at least one user who configures the controller to expect this error message. For the full disclosure, I don't have a controller code and never saw it, so I am not sure if that controller also handles the error of pulling error. But I can imagine there are plenty of implementations out the wild looking for this message as it is such a common and well-known behavior. So I am quite oppose of changing it, even though I understand how much better the experience you are suggesting is.

/cc @kubernetes/api-reviewers 

I wonder if API reviewers has an opinion here. @thockin @liggitt?

Reply:
values we put in REST API fields we say are intended to be programatically readable (like pod condition `Reason`, ContainerStateWaiting `Reason` fields, etc) are definitely integrated with as an API

Is this changing something that would make integrations like https://grep.app/search?q=ImagePullBackOff%22&case=true break?

Reply:
I totally see the point. Would it be possible to only change the `message`? The current implementation lets the `waiting` state still flip, but append the original error (SignatureValidationFailed) to the `reason`:

```json
"state": {
  "waiting": {
    "message": "Back-off pulling image \"quay.io/crio/unsigned:latest\": SignatureValidationFailed: image pull failed for quay.io/crio/unsigned:latest because the signature validation failed: Source image rejected: A signature was required, but no signature exists",
    "reason": "ImagePullBackOff"
  }
},
```

and

```json
"state": {
  "waiting": {
    "message": "image pull failed for quay.io/crio/unsigned:latest because the signature validation failed: Source image rejected: A signature was required, but no signature exists",
    "reason": "SignatureValidationFailed"
  }
},
```

Reply:
yeah, adjusting the human readable message is *way* less disruptive, in my opinion